// @dl-io/shared/src/schemas/ai-chat.ts
import { z } from "zod";

export const aiChatMessageSchema = z.object({
  role: z.enum(["system", "user", "assistant"]),
  content: z.string(),
});

export const aiChatRequestSchema = z.object({
  model: z.string().min(1),
  messages: z.array(aiChatMessageSchema).min(1),
  provider: z.enum(["openai", "anthropic", "google"]).optional(),
  temperature: z.number().min(0).max(2).optional(),
  max_tokens: z.number().int().min(1).max(128000).optional(),
  stream: z.boolean().optional().default(false),
  gateway_mode: z.enum(["byok", "pooled", "dedicated"]).optional(),
});

export const aiChatResponseSchema = z.object({
  id: z.string(),
  model: z.string(),
  provider: z.string(),
  choices: z.array(
    z.object({
      message: aiChatMessageSchema,
      finish_reason: z.string().nullable(),
    })
  ),
  usage: z.object({
    prompt_tokens: z.number(),
    completion_tokens: z.number(),
    total_tokens: z.number(),
  }),
});

export type AiChatRequest = z.infer<typeof aiChatRequestSchema>;
export type AiChatResponse = z.infer<typeof aiChatResponseSchema>;
