// @dl-io/shared/src/schemas/auth.ts
import { z } from "zod";

export const signupRequestSchema = z.object({
  email: z.string().email().max(255),
  password: z.string().min(8).max(128),
  name: z.string().min(1).max(100),
});

export const loginRequestSchema = z.object({
  email: z.string().email(),
  password: z.string().min(1),
});

export const authResponseSchema = z.object({
  token: z.string(),
  user: z.object({
    id: z.string(),
    email: z.string(),
    name: z.string(),
    plan: z.enum(["free", "pro", "enterprise"]),
  }),
});

export const createApiKeyRequestSchema = z.object({
  name: z.string().min(1).max(100),
});

export const apiKeyResponseSchema = z.object({
  id: z.string(),
  name: z.string(),
  key: z.string().optional(),
  maskedKey: z.string(),
  tier: z.string(),
  isActive: z.boolean(),
  createdAt: z.string(),
});

export type SignupRequest = z.infer<typeof signupRequestSchema>;
export type LoginRequest = z.infer<typeof loginRequestSchema>;
export type AuthResponse = z.infer<typeof authResponseSchema>;
export type CreateApiKeyRequest = z.infer<typeof createApiKeyRequestSchema>;
export type ApiKeyResponse = z.infer<typeof apiKeyResponseSchema>;

// Password reset
export const forgotPasswordSchema = z.object({ email: z.string().email() });
export const resetPasswordSchema = z.object({
  token: z.string().min(1),
  password: z.string().min(8).max(128),
});

// 2FA setup
export const enable2faResponseSchema = z.object({
  secret: z.string(),
  qr_uri: z.string(),
  backup_codes: z.array(z.string()),
});
export const verify2faSchema = z.object({ code: z.string().length(6) });
export const disable2faSchema = z.object({ code: z.string().length(6) });

// Login with 2FA
export const login2faSchema = z.object({
  temp_token: z.string(),
  code: z.string().min(6).max(8), // 6 for TOTP, 8 for backup code
});

// 2FA verify-setup (code + encrypted secret from setup step)
export const verifySetup2faSchema = z.object({
  code: z.string().length(6),
  encrypted_secret: z.string().min(1),
  encrypted_backup_codes: z.string().min(1),
});

export type ForgotPasswordRequest = z.infer<typeof forgotPasswordSchema>;
export type ResetPasswordRequest = z.infer<typeof resetPasswordSchema>;
export type Login2faRequest = z.infer<typeof login2faSchema>;
export type Verify2faRequest = z.infer<typeof verify2faSchema>;
export type Disable2faRequest = z.infer<typeof disable2faSchema>;
export type VerifySetup2faRequest = z.infer<typeof verifySetup2faSchema>;
