// @dl-io/shared/src/schemas/billing.ts
import { z } from "zod";

export const selectPlanRequestSchema = z.object({
  plan: z.enum(["free", "starter", "pro", "enterprise"]),
  modules: z.array(z.string()).min(1),
  teamId: z.string().optional(),
});

export const usageSummarySchema = z.object({
  plan: z.string(),
  billingPeriod: z.object({ start: z.string(), end: z.string() }),
  included: z.number(),
  used: z.number(),
  remaining: z.number(),
  overage: z.number(),
  overageCost: z.number(),
  byModule: z.array(
    z.object({
      module: z.string(),
      requests: z.number(),
      tokens: z.number(),
    })
  ),
});

export const onboardingRequestSchema = z.object({
  plan: z.enum(["free", "starter", "pro", "enterprise"]),
  modules: z.array(z.string()).min(1),
  teamName: z.string().optional(),
  teamSlug: z.string().optional(),
});

export type SelectPlanRequest = z.infer<typeof selectPlanRequestSchema>;
export type UsageSummary = z.infer<typeof usageSummarySchema>;
export type OnboardingRequest = z.infer<typeof onboardingRequestSchema>;
