// @dl-io/shared/src/schemas/gateway.ts
import { z } from "zod";

export const gatewayModeSchema = z.enum(["byok", "pooled", "dedicated"]);
export type GatewayMode = z.infer<typeof gatewayModeSchema>;

export const providerKeyCreateSchema = z.object({
  provider: z.enum(["openai", "anthropic", "google"]),
  apiKey: z.string().min(10),
  label: z.string().max(100).optional(),
  teamId: z.string().optional(),
});

export const providerKeyResponseSchema = z.object({
  id: z.string(),
  provider: z.string(),
  keyHint: z.string(),
  label: z.string(),
  isActive: z.boolean(),
  createdAt: z.string(),
});

export const creditBalanceResponseSchema = z.object({
  balanceCents: z.number(),
  lifetimePurchasedCents: z.number(),
  lifetimeUsedCents: z.number(),
  lastToppedUp: z.string().nullable(),
});

export const creditPurchaseRequestSchema = z.object({
  amountCents: z.number().int().min(500).max(1_000_000),
  teamId: z.string().optional(),
});

export const creditTransactionResponseSchema = z.object({
  id: z.string(),
  type: z.string(),
  amountCents: z.number(),
  balanceAfterCents: z.number(),
  description: z.string(),
  aiProvider: z.string().nullable(),
  model: z.string().nullable(),
  tokensUsed: z.number().nullable(),
  createdAt: z.string(),
});

export const tokenCostSchema = z.object({
  promptCostCents: z.number(),
  completionCostCents: z.number(),
  totalCostCents: z.number(),
});

export type TokenCost = z.infer<typeof tokenCostSchema>;

export const setGatewayModeSchema = z.object({
  mode: gatewayModeSchema,
  teamId: z.string().optional(),
});

export type ProviderKeyCreate = z.infer<typeof providerKeyCreateSchema>;
export type ProviderKeyResponse = z.infer<typeof providerKeyResponseSchema>;
export type CreditBalanceResponse = z.infer<typeof creditBalanceResponseSchema>;
export type CreditPurchaseRequest = z.infer<typeof creditPurchaseRequestSchema>;
export type CreditTransactionResponse = z.infer<typeof creditTransactionResponseSchema>;
export type SetGatewayMode = z.infer<typeof setGatewayModeSchema>;

export const autoReplenishSettingsSchema = z.object({
  enabled: z.boolean(),
  thresholdCents: z.number().int().min(100).max(10000),
  amountCents: z.number().int().min(500).max(50000),
});

export type AutoReplenishSettings = z.infer<typeof autoReplenishSettingsSchema>;
