// @dl-io/shared/src/schemas/html-sanitize.ts
import { z } from "zod";

export const htmlSanitizeRequestSchema = z.object({
  input: z.string().min(1).max(500_000),
  input_format: z.enum(["html", "markdown"]).optional().default("html"),
  output_format: z.enum(["html"]).optional().default("html"),
  allow_tags: z.array(z.string()).optional(),
  deny_tags: z.array(z.string()).optional(),
  allow_attributes: z.array(z.string()).optional(),
  deny_attributes: z.array(z.string()).optional(),
  strip_comments: z.boolean().optional().default(true),
});

export const htmlSanitizeResponseSchema = z.object({
  sanitized: z.string(),
  removed_count: z.number(),
  removed_items: z.array(
    z.object({
      type: z.enum([
        "tag",
        "attribute",
        "event_handler",
        "comment",
        "script",
      ]),
      name: z.string(),
      count: z.number(),
    })
  ),
});

export type HtmlSanitizeRequest = z.infer<typeof htmlSanitizeRequestSchema>;
export type HtmlSanitizeResponse = z.infer<typeof htmlSanitizeResponseSchema>;
