// @dl-io/shared/src/schemas/ip-intel.ts
import { z } from "zod";

export const ipIntelRequestSchema = z.object({
  ip: z.string().max(45),
  deep: z.boolean().optional().default(false),
});

export const ipIntelResponseSchema = z.object({
  ip: z.string(),
  geo: z.object({
    country: z.string().nullable(),
    country_code: z.string().nullable(),
    region: z.string().nullable(),
    city: z.string().nullable(),
    latitude: z.number().nullable(),
    longitude: z.number().nullable(),
    timezone: z.string().nullable(),
  }),
  network: z.object({
    asn: z.number().nullable(),
    isp: z.string().nullable(),
    org: z.string().nullable(),
  }),
  threat: z
    .object({
      is_vpn: z.boolean().nullable(),
      is_proxy: z.boolean().nullable(),
      is_tor: z.boolean().nullable(),
      is_datacenter: z.boolean().nullable(),
      risk_score: z.number().nullable(),
    })
    .nullable(),
});

export type IpIntelRequest = z.infer<typeof ipIntelRequestSchema>;
export type IpIntelResponse = z.infer<typeof ipIntelResponseSchema>;
