// @dl-io/shared/src/schemas/pdf-generate.ts
import { z } from "zod";

export const pdfGenerateRequestSchema = z
  .object({
    html: z.string().max(500_000).optional(),
    url: z.string().url().max(2048).optional(),
    pageSize: z
      .enum(["A4", "letter", "legal", "tabloid"])
      .optional()
      .default("A4"),
    margins: z
      .object({
        top: z.string().optional().default("1cm"),
        right: z.string().optional().default("1cm"),
        bottom: z.string().optional().default("1cm"),
        left: z.string().optional().default("1cm"),
      })
      .optional(),
    landscape: z.boolean().optional().default(false),
    headerTemplate: z.string().max(10_000).optional(),
    footerTemplate: z.string().max(10_000).optional(),
    printBackground: z.boolean().optional().default(true),
    waitFor: z.number().int().min(0).max(30000).optional().default(0),
  })
  .refine((data) => data.html || data.url, {
    message: "Either html or url must be provided",
  });

export const pdfGenerateResponseSchema = z.object({
  pdf: z.string(),
  pageCount: z.number(),
  sizeBytes: z.number(),
});

export type PdfGenerateRequest = z.infer<typeof pdfGenerateRequestSchema>;
export type PdfGenerateResponse = z.infer<typeof pdfGenerateResponseSchema>;
