// @dl-io/shared/src/schemas/teams.ts
import { z } from "zod";

export const createTeamRequestSchema = z.object({
  name: z.string().min(1).max(100),
  slug: z
    .string()
    .regex(/^[a-z0-9-]+$/, "Slug must be lowercase alphanumeric with hyphens")
    .min(3)
    .max(50),
});

export const inviteTeamMemberSchema = z.object({
  email: z.string().email(),
  role: z.enum(["admin", "member"]).default("member"),
});

export const updateTeamMemberRoleSchema = z.object({
  role: z.enum(["admin", "member"]),
});

export const teamResponseSchema = z.object({
  id: z.string(),
  name: z.string(),
  slug: z.string(),
  plan: z.string(),
  enabledModules: z.array(z.string()),
  memberCount: z.number(),
  createdAt: z.string(),
});

export const teamMemberResponseSchema = z.object({
  id: z.string(),
  userId: z.string(),
  name: z.string(),
  email: z.string(),
  role: z.string(),
  joinedAt: z.string(),
});

export type CreateTeamRequest = z.infer<typeof createTeamRequestSchema>;
export type InviteTeamMemberRequest = z.infer<typeof inviteTeamMemberSchema>;
export type TeamResponse = z.infer<typeof teamResponseSchema>;
export type TeamMemberResponse = z.infer<typeof teamMemberResponseSchema>;
