// @dl-io/shared/src/schemas/webhooks.ts
import { z } from "zod";

export const webhookEventTypes = [
  "moderation.flagged",
  "usage.threshold_80",
  "usage.threshold_90",
  "usage.threshold_100",
  "key.created",
  "key.revoked",
  "team.member_joined",
  "team.member_removed",
  "billing.payment_succeeded",
  "billing.payment_failed",
  "credits.low_balance",
  "credits.depleted",
] as const;

export const webhookEventTypeSchema = z.enum(webhookEventTypes);
export type WebhookEventType = z.infer<typeof webhookEventTypeSchema>;

export const createWebhookSchema = z.object({
  url: z.string().url().startsWith("https://"),
  events: z.array(webhookEventTypeSchema).min(1),
  description: z.string().max(200).optional(),
  teamId: z.string().optional(),
});

export const updateWebhookSchema = z.object({
  url: z.string().url().startsWith("https://").optional(),
  events: z.array(webhookEventTypeSchema).min(1).optional(),
  description: z.string().max(200).optional(),
  isActive: z.boolean().optional(),
});

export const webhookResponseSchema = z.object({
  id: z.string(),
  url: z.string(),
  secret: z.string(),
  events: z.array(z.string()),
  isActive: z.boolean(),
  description: z.string(),
  createdAt: z.string(),
});

export const webhookDeliveryResponseSchema = z.object({
  id: z.string(),
  event: z.string(),
  statusCode: z.number().nullable(),
  retryCount: z.number(),
  deliveredAt: z.string().nullable(),
  createdAt: z.string(),
});

export type CreateWebhookRequest = z.infer<typeof createWebhookSchema>;
export type UpdateWebhookRequest = z.infer<typeof updateWebhookSchema>;
export type WebhookResponse = z.infer<typeof webhookResponseSchema>;
export type WebhookDeliveryResponse = z.infer<typeof webhookDeliveryResponseSchema>;
