// @dl-io/shared/src/constants/ai-pricing.ts
// Costs in cents per 1,000 tokens
export const AI_MODEL_COSTS: Record<string, { promptPer1k: number; completionPer1k: number }> = {
  // OpenAI
  "gpt-4o": { promptPer1k: 0.25, completionPer1k: 1.0 },
  "gpt-4o-mini": { promptPer1k: 0.015, completionPer1k: 0.06 },
  "gpt-4-turbo": { promptPer1k: 1.0, completionPer1k: 3.0 },
  "o1": { promptPer1k: 1.5, completionPer1k: 6.0 },
  "o3-mini": { promptPer1k: 0.11, completionPer1k: 0.44 },
  "o4-mini": { promptPer1k: 0.11, completionPer1k: 0.44 },

  // Anthropic
  "claude-sonnet-4-5-20250929": { promptPer1k: 0.3, completionPer1k: 1.5 },
  "claude-opus-4-20250514": { promptPer1k: 1.5, completionPer1k: 7.5 },
  "claude-3-5-haiku-20241022": { promptPer1k: 0.08, completionPer1k: 0.4 },

  // Google
  "gemini-2.0-flash": { promptPer1k: 0.01, completionPer1k: 0.04 },
  "gemini-2.0-pro": { promptPer1k: 0.125, completionPer1k: 0.5 },
  "gemini-1.5-pro": { promptPer1k: 0.125, completionPer1k: 0.5 },
};

// Gateway mode pricing multipliers (applied to base provider token costs)
// Higher plans get lower rates — economies of scale
export const GATEWAY_MULTIPLIERS = {
  starter:    { pooled: 1.5, dedicated: 2.0 },
  pro:        { pooled: 1.3, dedicated: 1.7 },
  enterprise: { pooled: 1.1, dedicated: 1.4 },
} as const;

// Backwards-compat defaults (starter tier)
export const POOLED_MARKUP = GATEWAY_MULTIPLIERS.starter.pooled;
export const DEDICATED_MARKUP = GATEWAY_MULTIPLIERS.starter.dedicated;

export type GatewayPlanTier = keyof typeof GATEWAY_MULTIPLIERS;

export function getGatewayMultiplier(plan: string, mode: "pooled" | "dedicated"): number {
  const tier = plan in GATEWAY_MULTIPLIERS ? (plan as GatewayPlanTier) : "starter";
  return GATEWAY_MULTIPLIERS[tier][mode];
}

// Credit pack options
export const CREDIT_PACKS = [
  { id: "pack_500", amountCents: 500, label: "$5.00", bonusCents: 0 },
  { id: "pack_2000", amountCents: 2000, label: "$20.00", bonusCents: 100 },
  { id: "pack_5000", amountCents: 5000, label: "$50.00", bonusCents: 500 },
  { id: "pack_10000", amountCents: 10000, label: "$100.00", bonusCents: 1500 },
  { id: "pack_50000", amountCents: 50000, label: "$500.00", bonusCents: 10000 },
] as const;

// Stripe price IDs — populated by `npx tsx scripts/stripe-sync.ts`
// These are lookup_keys used to resolve prices at runtime.
// The sync script upserts products/prices and writes a manifest with real IDs.
export const STRIPE_PLAN_PRICES: Record<string, string> = {
  starter: "price_starter_monthly",
  starter_annual: "price_starter_annual",
  pro: "price_pro_monthly",
  pro_annual: "price_pro_annual",
  enterprise: "price_enterprise_monthly",
  enterprise_annual: "price_enterprise_annual",
};

// Stripe price IDs for credit packs — populated by stripe-sync.ts
export const STRIPE_CREDIT_PRICES: Record<string, string> = {
  pack_500: "price_credits_500",
  pack_2000: "price_credits_2000",
  pack_5000: "price_credits_5000",
  pack_10000: "price_credits_10000",
  pack_50000: "price_credits_50000",
};

// RevenueCat / App Store / Play Store product IDs
export const IAP_PRODUCT_IDS: Record<string, string> = {
  starter: "com.dickless.io.starter.monthly",
  starter_annual: "com.dickless.io.starter.annual",
  pro: "com.dickless.io.pro.monthly",
  pro_annual: "com.dickless.io.pro.annual",
  enterprise: "com.dickless.io.enterprise.monthly",
  enterprise_annual: "com.dickless.io.enterprise.annual",
};
