// @dl-io/shared/src/constants/pricing.ts
export const MODULES = {
  moderate: {
    id: "moderate",
    name: "Content Moderation",
    description: "Text & image toxicity detection",
    icon: "Shield",
    unitName: "scan",
  },
  redact: {
    id: "redact",
    name: "PII Redaction",
    description: "Strip personal data from text",
    icon: "EyeOff",
    unitName: "redaction",
  },
  ai_gateway: {
    id: "ai_gateway",
    name: "AI Gateway",
    description: "Unified proxy to OpenAI, Anthropic, Google",
    icon: "Zap",
    unitName: "request",
  },
  sanitize: {
    id: "sanitize",
    name: "Prompt Sanitizer",
    description: "Detect & strip prompt injections",
    icon: "ShieldAlert",
    unitName: "scan",
  },
  shorten: {
    id: "shorten",
    name: "URL Shortener",
    description: "Short links with analytics & QR codes",
    icon: "Link2",
    unitName: "link",
  },
  roast: {
    id: "roast",
    name: "Roast Tool",
    description: "AI-generated roasts for anything",
    icon: "Flame",
    unitName: "roast",
  },
  validate: {
    id: "validate",
    name: "Input Validation",
    description: "Email, phone, URL & domain validation",
    icon: "CheckCircle",
    unitName: "check",
  },
  ocr: {
    id: "ocr",
    name: "OCR / Text Extraction",
    description: "Extract text from images & PDFs",
    icon: "ScanText",
    unitName: "extraction",
  },
  translate: {
    id: "translate",
    name: "Translation",
    description: "Multi-language AI translation",
    icon: "Languages",
    unitName: "translation",
  },
  screenshot: {
    id: "screenshot",
    name: "Screenshot",
    description: "Capture any URL as PNG or PDF",
    icon: "Camera",
    unitName: "capture",
  },
  sentiment: {
    id: "sentiment",
    name: "Sentiment Analysis",
    description: "Detect sentiment & emotion in text",
    icon: "Heart",
    unitName: "analysis",
  },
  summarize: {
    id: "summarize",
    name: "Summarization",
    description: "Summarize text or web pages",
    icon: "FileText",
    unitName: "summary",
  },
  pdf_generate: {
    id: "pdf_generate",
    name: "PDF Generation",
    description: "Generate PDFs from HTML or URLs",
    icon: "FileDown",
    unitName: "generation",
  },
  email_verify: {
    id: "email_verify",
    name: "Email Verification",
    description: "MX lookup, SMTP check, disposable detection",
    icon: "MailCheck",
    unitName: "verification",
  },
  dns_lookup: {
    id: "dns_lookup",
    name: "DNS / WHOIS Lookup",
    description: "Query DNS records and WHOIS registrar data",
    icon: "Globe",
    unitName: "lookup",
  },
  ip_intel: {
    id: "ip_intel",
    name: "IP Geolocation & Threat Intel",
    description: "IP to location, ISP/ASN, VPN/proxy detection",
    icon: "MapPin",
    unitName: "lookup",
  },
  webhook_deliver: {
    id: "webhook_deliver",
    name: "Webhook Delivery",
    description: "Managed webhook delivery with retries and signing",
    icon: "Send",
    unitName: "delivery",
  },
  html_sanitize: {
    id: "html_sanitize",
    name: "HTML/Markdown Sanitizer",
    description: "XSS prevention, tag/attribute allow/deny lists",
    icon: "ShieldCheck",
    unitName: "sanitization",
  },
} as const;

export type ModuleId = keyof typeof MODULES;

export const ALL_MODULE_IDS: ModuleId[] = [
  "moderate",
  "redact",
  "ai_gateway",
  "sanitize",
  "shorten",
  "roast",
  "validate",
  "ocr",
  "translate",
  "screenshot",
  "sentiment",
  "summarize",
  "pdf_generate",
  "email_verify",
  "dns_lookup",
  "ip_intel",
  "webhook_deliver",
  "html_sanitize",
];

// Modules included in free & starter tiers (no AI-heavy or compute-heavy modules)
const FREE_MODULE_IDS: ModuleId[] = [
  "moderate",
  "redact",
  "sanitize",
  "shorten",
  "roast",
  "validate",
  "sentiment",
  "summarize",
  "email_verify",
  "dns_lookup",
  "ip_intel",
  "html_sanitize",
];

export const PLANS = {
  free: {
    id: "free",
    name: "Free",
    price: 0,
    priceLabel: "$0",
    period: "forever",
    description: "Kick the tires",
    requestsPerMinute: 10,
    monthlyRequests: 1_000,
    maxKeys: 2,
    maxTeamMembers: 0,
    monthlyCredits: 0,
    includedModules: FREE_MODULE_IDS as ModuleId[],
    features: [
      "1,000 requests/month",
      "10 req/min rate limit",
      "18 modules",
      "Community support",
      "2 API keys",
    ],
    overageRate: null as number | null, // no overage, hard cap
  },
  starter: {
    id: "starter",
    name: "Starter",
    price: 9,
    priceLabel: "$9",
    period: "/month",
    description: "For side projects & MVPs",
    requestsPerMinute: 30,
    monthlyRequests: 10_000,
    maxKeys: 5,
    maxTeamMembers: 0,
    monthlyCredits: 500,
    includedModules: FREE_MODULE_IDS as ModuleId[],
    features: [
      "10,000 requests/month",
      "30 req/min rate limit",
      "18 modules",
      "500 AI credits/mo",
      "Email support",
      "5 API keys",
      "$0.002 per extra request",
    ],
    overageRate: 0.002,
  },
  pro: {
    id: "pro",
    name: "Pro",
    price: 49,
    priceLabel: "$49",
    period: "/month",
    description: "For production apps",
    requestsPerMinute: 200,
    monthlyRequests: 100_000,
    maxKeys: 20,
    maxTeamMembers: 5,
    monthlyCredits: 1_500,
    includedModules: ALL_MODULE_IDS as ModuleId[],
    features: [
      "100,000 requests/month",
      "200 req/min rate limit",
      "All 18 modules (incl. AI Gateway, OCR, PDF, Webhooks)",
      "1,500 AI credits/mo",
      "Priority support",
      "20 API keys",
      "Team support (up to 5)",
      "$0.001 per extra request",
    ],
    overageRate: 0.001,
  },
  enterprise: {
    id: "enterprise",
    name: "Enterprise",
    price: 199,
    priceLabel: "$199",
    period: "/month",
    description: "For scale",
    requestsPerMinute: 1000,
    monthlyRequests: 1_000_000,
    maxKeys: 100,
    maxTeamMembers: 50,
    monthlyCredits: 7_500,
    includedModules: ALL_MODULE_IDS as ModuleId[],
    features: [
      "1,000,000 requests/month",
      "1,000 req/min rate limit",
      "All 18 modules",
      "7,500 AI credits/mo",
      "Dedicated support + SLA",
      "100 API keys",
      "Team support (up to 50)",
      "Bring your own AI keys + Webhook Delivery",
      "$0.0005 per extra request",
    ],
    overageRate: 0.0005,
  },
} as const;

export type PlanId = keyof typeof PLANS;

// Per-module pricing for metered add-ons
export const MODULE_PRICING: Record<
  ModuleId,
  { includedInFree: boolean; perRequestCost: number }
> = {
  moderate: { includedInFree: true, perRequestCost: 0.001 },
  redact: { includedInFree: true, perRequestCost: 0.001 },
  ai_gateway: { includedInFree: false, perRequestCost: 0.003 },
  sanitize: { includedInFree: true, perRequestCost: 0.001 },
  shorten: { includedInFree: true, perRequestCost: 0.0001 },
  roast: { includedInFree: true, perRequestCost: 0.002 },
  validate: { includedInFree: true, perRequestCost: 0.0005 },
  ocr: { includedInFree: false, perRequestCost: 0.005 },
  translate: { includedInFree: false, perRequestCost: 0.003 },
  screenshot: { includedInFree: false, perRequestCost: 0.01 },
  sentiment: { includedInFree: true, perRequestCost: 0.002 },
  summarize: { includedInFree: true, perRequestCost: 0.002 },
  pdf_generate: { includedInFree: false, perRequestCost: 0.01 },
  email_verify: { includedInFree: true, perRequestCost: 0.001 },
  dns_lookup: { includedInFree: true, perRequestCost: 0.0005 },
  ip_intel: { includedInFree: true, perRequestCost: 0.001 },
  webhook_deliver: { includedInFree: false, perRequestCost: 0.005 },
  html_sanitize: { includedInFree: true, perRequestCost: 0.0005 },
};

// ─── Annual Pricing ───────────────────────────────────────────────────

export const ANNUAL_DISCOUNT = 0.17; // ~17% off monthly equivalent

export const PLANS_ANNUAL = {
  starter: { price: 90, priceLabel: "$90", period: "/year", monthlySavings: 18 },
  pro: { price: 490, priceLabel: "$490", period: "/year", monthlySavings: 98 },
  enterprise: { price: 1990, priceLabel: "$1,990", period: "/year", monthlySavings: 398 },
} as const;

// ─── Trust & Safety Bundle ────────────────────────────────────────────

export const BUNDLES = {
  trust_and_safety: {
    id: "trust_and_safety",
    name: "Trust & Safety",
    description: "Content moderation + PII redaction + prompt sanitization + sentiment analysis at 25% discount",
    modules: ["moderate", "redact", "sanitize", "sentiment"] as ModuleId[],
    discountPercent: 25,
    price: 29,
    priceLabel: "$29",
    period: "/month",
  },
} as const;

// ─── Compliance Add-On ───────────────────────────────────────────────

export const COMPLIANCE_ADDON = {
  price: 99,
  priceLabel: "$99",
  period: "/month",
  features: [
    "Full audit logs for all API operations",
    "Data residency options (US, EU)",
    "Extended data retention (1 year)",
    "Dedicated compliance support",
    "SOC 2 readiness documentation",
  ],
} as const;

// ─── Per-Seat Team Pricing ───────────────────────────────────────────

export const SEAT_PRICING = {
  includedSeats: { free: 0, starter: 0, pro: 5, enterprise: 50 } as Record<string, number>,
  extraSeatPrice: 12, // $12/seat/month
} as const;

// ─── Helpers ──────────────────────────────────────────────────────────

export function isAiModule(moduleId: ModuleId): boolean {
  return moduleId === "ai_gateway";
}
