// @dl-io/shared/src/types/enums.ts
export const RoastType = {
  RESUME: "resume",
  LANDING_PAGE: "landing_page",
  CODE: "code",
  LINKEDIN: "linkedin",
  GENERAL: "general",
} as const;

export type RoastType = (typeof RoastType)[keyof typeof RoastType];

export const PlanTier = {
  FREE: "free",
  STARTER: "starter",
  PRO: "pro",
  ENTERPRISE: "enterprise",
} as const;

export type PlanTier = (typeof PlanTier)[keyof typeof PlanTier];

export const AIProvider = {
  OPENAI: "openai",
  ANTHROPIC: "anthropic",
  GOOGLE: "google",
} as const;

export type AIProvider = (typeof AIProvider)[keyof typeof AIProvider];

export const ModerationCategory = {
  VIOLENCE: "violence",
  HATE: "hate",
  SEXUAL: "sexual",
  SELF_HARM: "self_harm",
  DANGEROUS: "dangerous",
  HARASSMENT: "harassment",
} as const;

export type ModerationCategory =
  (typeof ModerationCategory)[keyof typeof ModerationCategory];

// GatewayMode type is exported from schemas/gateway.ts via Zod

export const CreditTransactionType = {
  PURCHASE: "purchase",
  USAGE: "usage",
  REFUND: "refund",
  BONUS: "bonus",
  PLAN_CREDIT: "plan_credit",
} as const;

export type CreditTransactionType =
  (typeof CreditTransactionType)[keyof typeof CreditTransactionType];

export const SubscriptionStatus = {
  NONE: "none",
  ACTIVE: "active",
  TRIALING: "trialing",
  PAST_DUE: "past_due",
  CANCELED: "canceled",
  UNPAID: "unpaid",
  INCOMPLETE_EXPIRED: "incomplete_expired",
} as const;

export type SubscriptionStatus =
  (typeof SubscriptionStatus)[keyof typeof SubscriptionStatus];

export const BillingInterval = {
  MONTH: "month",
  YEAR: "year",
} as const;

export type BillingInterval =
  (typeof BillingInterval)[keyof typeof BillingInterval];
